File: web/src/api/overdueCommentary.ts
/**
 * Overdue Commentary API
 */
import api from './client';

export interface UserResponse {
    user_id: number;
    email: string;
    full_name: string;
    role: string;
}

export interface OverdueComment {
    comment_id: number;
    validation_request_id: number;
    overdue_type: 'PRE_SUBMISSION' | 'VALIDATION_IN_PROGRESS';
    reason_comment: string;
    target_date: string;
    created_by_user_id: number;
    created_by_user: UserResponse;
    created_at: string;
    is_current: boolean;
    superseded_at: string | null;
    superseded_by_comment_id: number | null;
}

export interface CurrentOverdueCommentaryResponse {
    validation_request_id: number;
    model_id: number;
    model_name: string;
    overdue_type: 'PRE_SUBMISSION' | 'VALIDATION_IN_PROGRESS' | null;
    has_current_comment: boolean;
    current_comment: OverdueComment | null;
    is_stale: boolean;
    stale_reason: string | null;
    computed_completion_date: string | null;
}

export interface OverdueCommentaryHistoryResponse {
    validation_request_id: number;
    model_id: number;
    model_name: string;
    current_comment: OverdueComment | null;
    comment_history: OverdueComment[];
}

export interface CreateOverdueCommentaryRequest {
    overdue_type: 'PRE_SUBMISSION' | 'VALIDATION_IN_PROGRESS';
    reason_comment: string;
    target_date: string;
}

export interface MyOverdueItem {
    overdue_type: 'PRE_SUBMISSION' | 'VALIDATION_IN_PROGRESS';
    request_id: number;
    model_id: number;
    model_name: string;
    risk_tier: string | null;
    days_overdue: number;
    due_date: string | null;
    user_role: 'owner' | 'developer' | 'delegate' | 'validator';
    current_status: string;
    comment_status: 'CURRENT' | 'STALE' | 'MISSING';
    latest_comment: string | null;
    latest_comment_date: string | null;
    target_date: string | null;
    needs_comment_update: boolean;
}

export const overdueCommentaryApi = {
    /**
     * Get overdue commentary for a validation request
     */
    getForRequest: async (requestId: number): Promise<CurrentOverdueCommentaryResponse> => {
        const response = await api.get(`/validation-workflow/requests/${requestId}/overdue-commentary`);
        return response.data;
    },

    /**
     * Create or update overdue commentary for a validation request
     */
    createForRequest: async (requestId: number, data: CreateOverdueCommentaryRequest): Promise<OverdueComment> => {
        const response = await api.post(`/validation-workflow/requests/${requestId}/overdue-commentary`, data);
        return response.data;
    },

    /**
     * Get overdue commentary history for a validation request
     */
    getHistoryForRequest: async (requestId: number): Promise<OverdueCommentaryHistoryResponse> => {
        const response = await api.get(`/validation-workflow/requests/${requestId}/overdue-commentary/history`);
        return response.data;
    },

    /**
     * Get overdue commentary for a model (convenience endpoint)
     */
    getForModel: async (modelId: number): Promise<CurrentOverdueCommentaryResponse> => {
        const response = await api.get(`/models/${modelId}/overdue-commentary`);
        return response.data;
    },

    /**
     * Get current user's overdue items
     */
    getMyOverdueItems: async (): Promise<MyOverdueItem[]> => {
        const response = await api.get('/validation-workflow/dashboard/my-overdue-items');
        return response.data;
    }
};

export default overdueCommentaryApi;

File: api/app/schemas/overdue_commentary.py
"""Overdue Revalidation Commentary schemas."""
from pydantic import BaseModel, ConfigDict, Field
from datetime import datetime, date
from typing import Optional, Literal
from app.schemas.user import UserResponse


class OverdueCommentaryBase(BaseModel):
    """Base schema for overdue commentary."""
    reason_comment: str = Field(..., min_length=10, description="Explanation for the overdue status")
    target_date: date = Field(..., description="Target Submission Date (PRE_SUBMISSION) or Target Completion Date (VALIDATION_IN_PROGRESS)")


class OverdueCommentaryCreate(OverdueCommentaryBase):
    """Schema for creating overdue commentary."""
    overdue_type: Literal["PRE_SUBMISSION", "VALIDATION_IN_PROGRESS"] = Field(
        ..., description="Type of overdue: PRE_SUBMISSION or VALIDATION_IN_PROGRESS"
    )


class OverdueCommentaryResponse(OverdueCommentaryBase):
    """Response schema for a single overdue comment."""
    comment_id: int
    validation_request_id: int
    overdue_type: str
    created_by_user_id: int
    created_by_user: UserResponse
    created_at: datetime
    is_current: bool
    superseded_at: Optional[datetime] = None
    superseded_by_comment_id: Optional[int] = None

    model_config = ConfigDict(from_attributes=True, protected_namespaces=())


class CurrentOverdueCommentaryResponse(BaseModel):
    """Response schema for current overdue commentary status."""
    validation_request_id: int
    model_id: int
    model_name: str
    overdue_type: Optional[str] = None
    has_current_comment: bool
    current_comment: Optional[OverdueCommentaryResponse] = None
    is_stale: bool = False
    stale_reason: Optional[str] = None
    computed_completion_date: Optional[date] = None

    model_config = ConfigDict(from_attributes=True, protected_namespaces=())


class OverdueCommentaryHistoryResponse(BaseModel):
    """Response schema for overdue commentary history."""
    validation_request_id: int
    model_id: int
    model_name: str
    current_comment: Optional[OverdueCommentaryResponse] = None
    comment_history: list[OverdueCommentaryResponse] = []

    model_config = ConfigDict(from_attributes=True, protected_namespaces=())
